architecture/2024-11-ucp-component-threat-model.md

```diff
@@ -0,0 +1,247 @@
+      _Note:_ As previously noted, SHA-1 is used for optimization purposes only.
+
+4. **Hashing Input Data to Generate New ETags**: [Link to code](https://github.com/radius-project/radius/blob/main/pkg/ucp/util/etag/etag.go#L30)
+
+   1. **Purpose:** To hash input data to generate new ETags for resource versioning.
+   2. **Library:** Uses the Go standard `crypto/sha1` package: [crypto/sha1](https://pkg.go.dev/crypto/sha1)
+   3. **Type:** SHA-1  
+      _Note:_ SHA-1 usage here is for generating ETags and not for cryptographic security.
+
+---
+
+##### General Note on SHA-1 Usage
+
+In all instances where SHA-1 is utilized within the codebase, it serves for generating hashes to ensure uniqueness or detect changes efficiently. It is **not** used for cryptographic security purposes due to known vulnerabilities. For security-critical hashing, a more secure algorithm (e.g., SHA-256) should be employed.
+
+---
+
+#### Storage of Secrets
+
+Below you will find where and how Radius stores secrets. We create Kubernetes Secret objects and rely on Kubernetes security measures to protect these secrets.
+
+1. **Creating or Updating a Cloud Provider Credential Resource**: [Azure](https://github.com/radius-project/radius/blob/95409fe179d7adca884a3fc1d82f326bc81c8da0/pkg/ucp/frontend/controller/credentials/azure/createorupdateazurecredential.go#L89), [AWS](https://github.com/radius-project/radius/blob/95409fe179d7adca884a3fc1d82f326bc81c8da0/pkg/ucp/frontend/controller/credentials/aws/createorupdateawscredential.go#L81). When a **Cloud Provider Credential** resource is deployed, Radius creates a new Kubernetes Secret or updates an existing one. The Kubernetes Secret is deleted when the **Cloud Provider Credential** is deleted.
+
+#### Data Serialization / Formats
+
+We use custom parsers to parse Radius-related resource IDs and do not use any other custom parsers and instead rely on Kubernetes built-in parsers. Therefore, we trust Kubernetes security measures to handle data serialization and formats securely. The custom parser that parses Radius resource IDs has its own security mechanisms that don't accept anything other than a Radius resource ID.
+
+### Clients
+
+In this section, we will discuss the different clients of the Radius UCP (Universal Control Plane) component. Clients are systems that interact with the UCP component to trigger actions. Here are the clients of the UCP component:
+
+1. **Radius CLI**: Radius CLI interacts with UCP under the hood through the generated clients. For example, a `rad deploy` command needs to go to UCP and then through UCP to the Deployment Engine. UCP will do the forwarding of the requests to the Deployment Engine and the other components like Applications RP.
+
+2. **Direct API Querying**: Clients can interact with the UCP component directly by making API queries. For example, a client can query the status of a deployment or retrieve information about a specific resource by making a direct API call to the UCP server.
```

Comment:
**Every** component of Radius is a client of UCP 😆. I think we can simplify this section. There's a lot of text here, but it's not conveying a ton of information.

Reply:
Here's a question that **is** worth answering. How do these clients authenticate?

Reply:
Is Applications-RP a client of UCP? By client, I mean, is Applications-RP making a call to UCP? Or is UCP just forwarding the requests to Applications-RP?

Reply:
>  Is Applications-RP a client of UCP? By client, I mean, is Applications-RP making a call to UCP? Or is UCP just forwarding the requests to Applications-RP?

Both. We should assume that every RP is both client and destination.